---
title: The impact of ancestral, genetic, and environmental influences on germline
  de novo mutation rates and spectra
author: "O. Isaac Garcia-Salinas"
date: "2025-03-19"
output: 
  html_document: 
    fig_caption: yes
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Run .libPaths() pointing to a valid library directory before knitting!

```{r packages, message=FALSE, warning=FALSE}

.libPaths(c(.libPaths(), "/tools/aws-workspace-apps/ce/R/4.2.1")) # Valid path to libraries in GEL as of 20/03/2025

# Ensure the next libraries and versions can be loaded
library(vroom)
library(tidyverse)
library(MicrobiomeStat)
library(ggsignif)
library(ggrepel)
library(knitr)

sessionInfo()

```

Here, we present code that will allow users to reproduce the main analyses in our paper "The impact of ancestral, genetic, and environmental influences on germline de novo mutation rates and spectra" (Garcia-Salinas et al.,2025). This code and the datasets mentioned trhoughout are available inside GEL's reasearch environment in under the path: /re_gecip/shared_allGeCIPs/aeg_dnm_2024.

Access to the research environment is possible by following the instructions in <https://www.genomicsengland.co.uk/join-us>.

# Analyses on ancestry associations with DNM rate and spectra

## Dataset description

The next dataset contains the necessary data to run **Model 1** and **Model 2** described in our **main Methods**. This will allow users to reproduce our **main Figure 1**. The presented dataset was obatained by parsing three GEL tables (available through labkey), namely: a) denovo_cohort_information, b) aggregate_gvcf_sample_stats, and c) denovo_flagged_variants.

Genetically inferred ancestry was obtained from: /gel_data_resources/main_programme/aggregation/aggregate_gVCF_strelka/aggV2/additional_data/ancestry/MAF5_superPop_predicted_ancestries.tsv

```{r trio metadata, echo=FALSE, message=FALSE, warning=FALSE}

# Path to minimal dataset - non admixed trios 9820 rows x 31 columns
working_dir = "/nas/weka.gel.zone/re_gecip/shared_allGeCIPs/aeg_dnm_2024" ## Assuming access from the RE desktop, not the double helix HPC!

# Ensure setting column data types to avoid complications later
trio_metadata_non_admixed = vroom(
  file.path(working_dir, "non_admixed_trios_dnm_plus_metadata.tsv"), 
  delim="\t", 
  col_types = "ccccccccccccciiiddddddddddddddd") %>%
  mutate(trio_ancestry=factor(trio_ancestry, 
                              levels=c("afr", "amr", "eas", "eur", "sas")))

```

Including metadata for 9820 trios.

```{r trio count report}

# Each row is a trio
nrow(trio_metadata_non_admixed)

#N trios per ancestry
table(trio_metadata_non_admixed$trio_ancestry)

```

The provided minimal dataset includes trio IDs, participant IDs, and plate key IDs which can be used to find and merge different metadata available in the GEL ecosystem.

MEMBER, One of: Father, Mother, Offspring. PYR_SUBS, One of: C2A, C2G, C2T, CpG2TpG, T2A, T2C, T2G.

```{r trio metadata column description, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}

trio_dataset_cols = "
| column_name | data type | content |
|-------------|:----------|:--------|
|trio_id | character | GEL trio IDs, including only one trio per family |
| family_id | character | GEL family IDs, can be linked to multiple trios |
| participant_id_TRIO_MEMBER  | character | GEL participant ID for the TRIO MEMBER | 
| plate_key_TRIO_MEMBER | character GEL plate key ID for the TRIO MEMBER (use to find BAM files) |
| plate_key_TRIO_MEMBER | character GEL plate key ID for the TRIO MEMBER (use to find BAM files) |
| ancestry_MEMBER | character | GEL assigned ancestry for individual according to https://re-docs.genomicsengland.co.uk/ancestry_inference/ |
| is.admixed_offspring | character | Flag (0|1) to identify offspring of individuals with discordant ancestries (1) or from individuals with the same predicted ancestry (0) | 
| trio_ancestry | factor | One of: afr, eas, amr, eur, sas | 
| n.dnSNVs | integer | total de novo SNVs detected in trio |
| age_at_offs_Mother | integer | Maternal age at conception | 
| age_at_offs_Father | integer | Paternal age at conception | 
| illumina_mean_coverage_MEMBER| double | Mean WGS coverage as defined in the aggregate_gvcf_sample_stats table | 
| illumina_percent_aligned_reads_MEMBER | double | Percent aligned reads as defined in the aggregate_gvcf_sample_stats | 
| trio.med.bf | double | Median bayes factor (from platypus variant caller) across all DNMs called per trio | 
| trio.med.vaf  | double | Median VAF across all DNMs called per trio |
| prop_PYR_SUBS | double | Proportion of de novo SNVs classified as either PYR_SUBS out of the total de novo SNVs per trio |
"

cat(trio_dataset_cols)

```

## **Ancestry-associated differences on DNM rate**

Use the next chunk to run pairwise ancestry effect comparisons on DNM rate.

```{r dnm rate x ancestry, message=FALSE, warning=FALSE, results='asis'}

# The next function performs glm "quasipoisson" regression using the "test_formula" while changing the base ancestry one of afr, amr, eas, eur, or sas. 
# It then, extracts summary statistics for the effect of ancestry A compared to ancestry B on DNM rate 
# Estimated effects are in log(fold change) scale
# Remember to provide the ancestry colunmn index to know which ancestry colunmn to take (out of trio_ancestry, ancestry_father, etc).
# Here I identify the index using grep in a vector of column names
ancestry_level_refactor = function(ancestry_index, assoc_dataframe, test_formula, 
                                   ancestry_col_index){
  # base == first ancestry in the ancetry factor
  all_ancestry_level_combinations = list("eur" = c("eur", "amr", "eas", "afr", "sas"), 
     "amr" = c("amr", "eur", "eas", "afr", "sas"), 
     "eas" = c("eas", "eur", "amr", "afr", "sas"), 
     "afr" = c("afr", "eur", "amr", "eas", "sas"), 
     "sas" = c("sas", "eur", "amr", "eas", "afr"))
  
  # Use index to select the ancestyr list to use for level refactor
  ancestry_comparison = names(all_ancestry_level_combinations[ancestry_index]) ## A.k.a base ancestry
  curr_levels = all_ancestry_level_combinations[[ancestry_index]]
  
  message("Base ancestry is: ", ancestry_comparison)
  
  original_ancestry_name = colnames(assoc_dataframe)[[ancestry_col_index]] 
  
  # rename ancestry column for ease of coefficiente extraction
  colnames(assoc_dataframe)[ancestry_col_index] = "ancestry"
  
  # re-define base ancestry (first ancestry in list is base)
  dataset_reorder = assoc_dataframe %>%
    mutate(ancestry=factor(ancestry, levels=curr_levels))
  
  colnames(dataset_reorder)[ancestry_col_index] = original_ancestry_name
  
  message("Running quasipoisson family regression...")
  message("Formula is: ", 
            test_formula)
  curr_model = glm(formula=test_formula,
                   data=dataset_reorder, 
                   family="quasipoisson")
  
  curr_summary = summary(curr_model)
  
  # Extract ancestry effect coefficients  
  final_coefficients = curr_summary[["coefficients"]] %>%
    as.data.frame() %>%
    rownames_to_column(var ="vs") %>% 
    dplyr::slice(2:5) %>% # each row is a comparison ancestry A / ancestry B effect
    mutate(base=ancestry_comparison, 
           vs=gsub(vs, pattern=original_ancestry_name, replacement=""))
    
  colnames(final_coefficients) = c("vs", "logfc", "stderr", "tval", "nominal_p", "base")
  final_coefficients = final_coefficients[c(1,6,2:5)] %>%
    mutate(base=toupper(base), vs=toupper(vs))
    
  return(final_coefficients)
  
}

# Formula corrsponding to Model 1
qc_covars = paste(colnames(trio_metadata_non_admixed)[c(17:24)], collapse = "+")
dnSNV_rate_by_ancestry = paste0("n.dnSNVs~trio_ancestry+age_at_offs_Mother+age_at_offs_Father+", qc_covars)


# Loop through all possible ancesrtry bases (5)
dnSNVrate_byAncestry_allCombs = lapply(c(1:5), ancestry_level_refactor, 
                                       assoc_dataframe=trio_metadata_non_admixed, 
                                       test_formula=dnSNV_rate_by_ancestry, 
                                       ancestry_col_index=grep(colnames(trio_metadata_non_admixed),
                                                               pattern="trio_ancestry")) %>%
  purrr::reduce(rbind)

# Remove redundant ancestry comparisons

ancestry_combs_mock = data.frame(
  vs=combn(unique(dnSNVrate_byAncestry_allCombs[["vs"]]), m = 2)[1,],
  base=combn(unique(dnSNVrate_byAncestry_allCombs[["vs"]]), m = 2)[2,])

ancestry_combs_mock[10,] = c("EUR","SAS")
ancestry_combs_mock[5,] = c("AFR","EAS")
ancestry_combs_mock[2,] = c("AFR","AMR")
  
# The following table corresponds to supplementary data 1 
dnSNVrate_byAncestry_nonredundant = ancestry_combs_mock %>%
  left_join(., dnSNVrate_byAncestry_allCombs, by=c("vs", "base")) %>%
  mutate(padj=p.adjust(nominal_p, method = "fdr"),
         fdr_label=ifelse(padj<=0.05, y="*", n=""),
         fc=exp(logfc), 
         lwr=exp(logfc-1.96*stderr), 
         upr=exp(logfc+1.96*stderr),
         ancestry_pair=paste0(vs, "/", base))

```

See summary statistics corresponding to Supplementary data 1 (dnSNVrate_byAncestry_nonredundant).

```{r print supp data 1, echo=FALSE}

kable(dnSNVrate_byAncestry_nonredundant)

```

Plot non-redundant ancestry pair fold change differences on DNM rate

```{r plot ancestry x DNM rate fold changes, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}

dnSNVrate_byAncestry_nonredundant %>%
  ggplot(., aes(x=fc,y=ancestry_pair)) + 
    facet_wrap(~ancestry_pair, ncol=1, scales="free_y") + 
    geom_errorbarh(aes(xmin=lwr, xmax=upr, height=0), linewidth=6, color="#99B898") + 
    geom_vline(xintercept = 1, linetype="dotdash") + 
    geom_point(color="black", size=8) + 
    geom_text(aes(x=ifelse(fc<1, y=lwr-0.01, n=upr+0.01), 
                label=fdr_label), size=16, color="black", show.legend = F, 
            vjust=0.7, fontface="bold") + 
    labs(x="Fold change", y="Ancestry pair") + 
    theme_classic(base_size = 20) + 
    theme(strip.text.x = element_blank(), 
          axis.text = element_text(face="bold", size=14),
          strip.background = element_blank(),
          legend.position = "bottom", 
          legend.box.just = "left", 
          axis.title = element_text(face="bold", size=16), 
          legend.title=element_text(face="bold")) 


```

Alternatively, look at the expected baseline DNM counts per ancestry by looking at the intercept estimate from **Model 1**.

```{r get baseline DNM counts, message=FALSE, warning=FALSE, results='asis'}

# Identify numeric covariates to scale 
numeric_covariates = c("age_at_offs_Father", "age_at_offs_Mother" , 
               "illumina_mean_coverage_Father", "illumina_mean_coverage_Mother",
               "illumina_mean_coverage_Offspring",
               "illumina_percent_aligned_reads_Father",
               "illumina_percent_aligned_reads_Mother",
               "illumina_percent_aligned_reads_Offspring","trio.med.bf","trio.med.vaf")

numeric_covars_indices = sort(unlist(lapply(numeric_covariates, function(covar){
  index=grep(x=colnames(trio_metadata_non_admixed), pattern=paste0("\\b", covar, "\\b"))
  return(index)
}) ))

# reduce dataset to numeric covariates, trio_id, trio_ancestry, and modelled variable (n.dnSNVs)
trio_metadata_non_admixed_scaled = trio_metadata_non_admixed[,c(1,13,14, numeric_covars_indices)] 
# scale all numeric covariates
for(i in c(4:length(trio_metadata_non_admixed_scaled))){
  trio_metadata_non_admixed_scaled[,i] = scale(trio_metadata_non_admixed_scaled[,i])
}

# Run dnm rate x ancestry regression on scaled covariates
# The intercept of the regression will correspond to the log(nDNMs) expected as
# baseline for the ancestry used as the baseline 
# This function gets baseline_dnms and 95% confidence intervals
get_baseline_estimate = function(base_ancestry, 
                                 regression_df, 
                                 current_formula=NULL){
  
  all_ancestries = unique(regression_df[["trio_ancestry"]])
  non_base_ancestries = grep(all_ancestries, pattern = base_ancestry, invert = T,value = T)
  regression_df[["trio_ancestry"]] = factor(regression_df[["trio_ancestry"]], 
                                            levels=c(base_ancestry, non_base_ancestries))
  
  curr_regression = summary(glm(family="quasipoisson", 
                        formula = current_formula, 
                        data=regression_df))
  
  # Extract summary statistics for the model intercept - estimate effect for intercept is in log scale!
  intercepts = curr_regression[["coefficients"]][1, ]
  out_estimates = data.frame(estimate=intercepts[[1]], stderr=intercepts[[2]], 
                             tval=intercepts[[3]], p=intercepts[[4]], 
                             base_ancestry = base_ancestry) %>%
    mutate(baseline_dnms=exp(estimate), 
           lci=exp(estimate - stderr*1.96), 
           uci=exp(estimate + stderr*1.96))
  
  return(out_estimates)
  
}

# Estimate baseline DNMs while for each ancestry 
# (i.e. change model base for afr, amr, etc. each time)

# Formula is still dnSNV_rate_by_ancestry

baseline_esimates = lapply(c("afr", "amr", "eas", "eur", "sas"), 
                           get_baseline_estimate, current_formula= dnSNV_rate_by_ancestry, 
                           regression_df=trio_metadata_non_admixed_scaled) %>%
  purrr::reduce(rbind) %>%
  mutate(base_ancestry=toupper(base_ancestry))

```

See summary statistics corresponding to baseline DNM counts per ancestry (baseline_esimates).

```{r print baseline dnm counts, echo=FALSE}

kable(baseline_esimates)

```

Plot **main Figure 1A** corresponding to baseline DNM count expected for each ancestry

```{r plot baseline DNM counts, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}

# Significant pairs correspond to significantly different pairs from Model 1!
baseline_esimates %>%
  ggplot(., aes(x=base_ancestry, y=baseline_dnms)) + 
  geom_errorbar(aes(ymin=lci, ymax=uci), width=0, linewidth=6, color="#99B898") + 
  geom_point(size=8) + 
  labs(y="Baseline DNM count", x="Ancestry group") + 
    geom_signif(
    y_position = c(68.5,69.3,70.1), xmin = c(1,1,1), xmax = c(2,4,5),
    annotation = c("**", "**", "**"), tip_length = 0.01, textsize = 8) + 
  theme_classic(base_size = 20) + 
  theme(axis.title = element_text(face = "bold", size=16), 
        axis.text = element_text(face = "bold", size=14)) + 
  scale_y_continuous(limits = c(62,70.5))

```

## **Ancestry-associated differences on DNM spectra**

We modelled changes in DNM mutation spectra using a compositional linear regression model. We used the linda() function from the "MicrobiomeStat" package (Zhou et al., 2022).

```{r dnm spectra x ancestry using linda, message=FALSE, warning=FALSE, results='asis'}

#current_subset = trio_metadata_non_admixed

# We followed conventional formatting required by the linda function
# Get a pyrimidine substitution proportion matrix (trio x pyr subs)
pyr_subs_matrix = trio_metadata_non_admixed[, c(grep(colnames(trio_metadata_non_admixed), pattern="prop"))]
# Transpose matrix (7 x 9820)
pyr_subs_matrix_transposed = as.data.frame(t(pyr_subs_matrix))
colnames(pyr_subs_matrix_transposed) = trio_metadata_non_admixed[["trio_id"]]

# Identify numeric covariates to scale 
numeric_covariates = c("age_at_offs_Father", "age_at_offs_Mother" , 
               "illumina_mean_coverage_Father", "illumina_mean_coverage_Mother",
               "illumina_mean_coverage_Offspring",
               "illumina_percent_aligned_reads_Father",
               "illumina_percent_aligned_reads_Mother",
               "illumina_percent_aligned_reads_Offspring","trio.med.bf","trio.med.vaf")

numeric_covars_indices = sort(unlist(lapply(numeric_covariates, function(covar){
  index=grep(x=colnames(trio_metadata_non_admixed), pattern=paste0("\\b", covar, "\\b"))
  return(index)
}) ))

# reduce dataset to numeric covariates, trio_id and trio_ancestry - output variable not included as this is  "pyr_subs_matrix_transposed"
trio_metadata_non_admixed_scaled = trio_metadata_non_admixed[,c(1,13, numeric_covars_indices)] 
# scale all numeric covariates
for(i in c(3:length(trio_metadata_non_admixed_scaled))){
  trio_metadata_non_admixed_scaled[,i] = scale(trio_metadata_non_admixed_scaled[,i])
}

# Run a linda regression changing the model baseline each time
# linda estimates the effect of covariates on the whole proportions matrix so the formula excludes output
linda_formula = paste0("~trio_ancestry" , "+", paste(numeric_covariates, collapse = "+")) # This is Model 2 

# returns a list of linda objects - 1 per base ancestry
all_ancestry_linda_models = lapply(1:5, function(index){
  
  # Ordering of ancestry levels according to base ancestry
  ancestry_levels = list(c("sas", "afr", "amr", "eur", "eas"),
                       c("afr", "eur", "sas", "amr", "eas"),
                       c("eur", "sas", "afr", "amr", "eas"), 
                       c("eas", "sas", "afr", "amr", "eur"), 
                       c("amr", "sas", "afr", "eas", "eur"))

  names(ancestry_levels) = c("sas", "afr", "eur", "eas", "amr")
  
  curr_base_ancestry = ancestry_levels[[index]]
  base_ancestry_name = names(ancestry_levels[index])
  message("Running linda with base ancestry ", base_ancestry_name)
  
  # Re-define baseline of factor variable
  curr_base_ancestry_scaled = trio_metadata_non_admixed_scaled %>%
  mutate(trio_ancestry=factor(trio_ancestry, levels=curr_base_ancestry))
  
  current_linda = linda(pyr_subs_matrix_transposed, curr_base_ancestry_scaled, 
                  formula = linda_formula, 
                  feature.dat.type = "proportion", 
           prev.filter = 0.1, is.winsor = TRUE, outlier.pct = 0.03,
           p.adj.method = "fdr", alpha = 0.05) # fdr correction from linda apples to all covariates! Not used
  return(current_linda)
})

# Name corresponds to ancestry used as baseline
names(all_ancestry_linda_models) = c("sas", "afr", "eur", "eas", "amr")

## The next function parses the a single linda object to extract coefficient estimates corresponding to ancestry pairwise comparisons 

get_pairwise_effects = function(ancestry_pair, lindaObjects){
  base_ancestry = ancestry_pair[[2]]
  other_ancestry = paste0("trio_ancestry", ancestry_pair[[1]])
  #message(base_ancestry)
  #message(other_ancestry)
  other_ancestry_coeffs = lindaObjects[[base_ancestry]][["output"]][[other_ancestry]][, c("log2FoldChange","lfcSE", "stat", "pvalue")] %>%
    rownames_to_column(var="pyr") %>%
    mutate(baseAncestry=toupper(base_ancestry), otherAncestry=toupper(ancestry_pair[[1]]), 
           ancestry_pair=paste0(otherAncestry, "/", baseAncestry)) 
    #message(paste0(c(ancestry_pair[[2]], base_ancestry)) )
  return(other_ancestry_coeffs)
  
} 

# Use a list to exract non-redundant pairwise comparisons
all_ancestry_pairs = list(c("AFR","AMR"), c("AFR","EAS"), 
                          c("AFR","EUR"), c("AFR","SAS"), 
                          c("AMR","EAS"), c("AMR","EUR"), 
                          c("AMR","SAS"), c("EAS", "EUR"), 
                          c("EAS","SAS"), c("EUR","SAS"))

# Generate a single dataframe with all pairwise ancestry comparisons 
all_pairwise_linda_comparisons = lapply(lapply(all_ancestry_pairs, tolower), get_pairwise_effects, 
       lindaObjects=all_ancestry_linda_models) %>%
  purrr::reduce(rbind) %>%
  mutate(padj=p.adjust(pvalue,method="fdr"), 
         sig_lab=case_when(padj<=0.05 ~ "**", 
                           padj<=0.1 ~ "*", 
                           TRUE ~ ""),
         pyr=gsub(pyr, pattern="prop_", replacement=""), 
         pyr=gsub(pyr, pattern="2", replacement=">")) 

```

See summary statistics corresponding to Supplementary Data 2 (all_pairwise_linda_comparisons)

```{r print supp data 2, echo=FALSE}
kable(all_pairwise_linda_comparisons)
```

Plot **main Figure 1B** corresponding to pyrimidine substitution proportion fold change differences between ancestry pairs.

```{r plot ancestry x pyr subs fold changes, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}

# Just a helper function to get the highest raw pval that would pass FDR at different levels 
get_fdr_threshold= function(pvals, threshold=0.05){
  # Total number of p-values
  N = length(pvals)
  # Sort p-values in increasing order
  p_values_sorted = sort(pvals)
  # Calculate the rank-based thresholds
  thresholds = (1:N) / N * threshold
  # Find the largest p-value that satisfies the BH condition
  bh_condition = p_values_sorted <= thresholds
  pvalue_threshold = max(p_values_sorted[bh_condition])
  return(pvalue_threshold)
}

raw_pval_thresholdFDR5p = get_fdr_threshold(all_pairwise_linda_comparisons$pvalue, 
                                            threshold = 0.05)

raw_pval_thresholdFDR10p = get_fdr_threshold(all_pairwise_linda_comparisons$pvalue, 
                                            threshold = 0.1)

# Reproduce main figure 1 panel B

all_pairwise_linda_comparisons %>%
  mutate(sig_lab=ifelse(padj<= 0.1, y=ancestry_pair, n="")) %>%
  ggplot(., aes(x=log2FoldChange, y=-log10(pvalue), fill=pyr)) + 
  geom_point(pch=21, color= "black", size=3) + 
  geom_hline(yintercept = round(-log10(raw_pval_thresholdFDR5p), digits = 2), linetype="dotdash", color="red") + 
  geom_hline(yintercept = round(-log10(raw_pval_thresholdFDR10p), digits = 2), linetype="dotdash", color="orange") +
  geom_vline(xintercept = 0, linetype="dotdash", color="grey") +
  annotate("text", label="FDR 5%", x = -0.35, y=6.2,
           size=3, fontface="bold") + 
  annotate("text", label="FDR 10%", x = -0.35, y=2.45,
           size=3, fontface="bold") + 
  geom_text_repel(aes(label=sig_lab), 
                  nudge_x = -.05, 
    box.padding = 0.1,
    nudge_y = 0.5,
    segment.angle = 20, size=4, fontface="bold") +
  scale_x_continuous(limits = c(-0.4, 0.4)) +
  labs(x="log2(Fold change)", y="-log10(p value)",fill= "Pyrimidine\nsubstitution") + 
  theme_classic(base_size = 18) + 
  theme(axis.text = element_text(face="bold",  size=15), 
        axis.title = element_text(face="bold", size=16), 
        legend.title = element_text(face="bold", size=16), 
        legend.text = element_text(face="bold")) + 
  scale_fill_manual(
    values=c("C>A"="#b3de69", "C>G"="#fdb462", "CpG>TpG"= "#bebada", 
             "C>T"="#fccde5", "T>A"="#80b1d3", "T>C"="#fb8072", 
              "T>G"="#8dd3c7"))


```

# **Modelling individual smoking behaviour on phased DNM rate**

## Dataset description

Now we present code and datasets necesary to reproduce our associations with smoking behavior. We provide two minimal datasets

a)  "inds_wicd10": Contains informaton on individual plate_keys flagged by suspected smoking behaviour based on ICD10 codes. This informaton was parsed from GEL's "hes_apc" table accesible trhough LabKey.
b)  "phased_dnsnvs_long": Contains information on phased DNM mutation for each parent and the same metadata as the trio dataset. Note that all parents were included regardless of their ancestry classification. Phased DNMs per individual were obtained as described by Kaplanis et al., 2022.

```{r phased DNMs and smoking metadata, echo=FALSE}

inds_wicd10 = vroom(
  file.path(working_dir, "all_parents_w_icd10_smk_flagged.tsv"), 
  delim="\t", col_types = "cciii") ## Has less individuals than the GWAS-like parents table as this includes parents WITH at least 1 ICD10 entry!!

## The next dataframe contains individuals from all ancestries!

phased_dnsnvs_long = vroom(file.path(working_dir, "gwas_like_table_phased_dnSNVs_all_ancestries_updatedCounts.tsv"), 
                           delim="\t", col_types = "cciicdddddddddd") 

# Merge with individual smoking behavior information; encode factors
# Keep only individuals with APC data!
phased_dnsnvs_wicd10_flags = phased_dnsnvs_long %>%
  filter(plate_key %in% inds_wicd10[["plate_key"]]) %>%
  left_join(., inds_wicd10, by="plate_key") %>%
  mutate(sex=factor(sex, levels=c(0,1)), 
         ever_smoked=factor(ever_smoked, levels=c(0,1)))

```

After merging information in "inds_wicd10" and "phased_dnsnvs_long", the working dataset for this section (phased_dnsnvs_wicd10_flags) contains metadata columns that match those described in the trio metadata and the followning new columns:

```{r smoking dataset column description, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}

smoking_dataset_cols = "
| column_name | data type | content |
|-------------|:----------|:--------|
| plate_key | character | Character GEL plate key ID for the individual (parent) in question |
| age_at_offs | interger | Individual age at conception | 
| phased_dnsnvs | integer | N phased DNMs to this individual | 
| ancestry | factor | Genetically inferred ancestry for this individual | 
| sex | factor | Binary flag (0|1) for individual sex (i.e., parent). Either (0) for fathers, or (1) for mothers | 
| illumina_snvs_MEMBER | integer | N SNVs for individual MEMBER as defined in the aggregate_gvcf_sample_stats |
| member | character | Trio member. One of: Mother, Father | 
| has.icd_entry | factor | Binary flag (0|1) indicating if individual had any ICD10 records (hes_apc) available (1) | 
| ever_smoked | factor | Binary flag (0|1) indicating if individual had either F17 or Z58.7 records (1) among their ICD10 records |
"
cat(smoking_dataset_cols)

```

Including metadata for 15,732 individuals with available APC records.

```{r n individuals in this dataset}

table(phased_dnsnvs_wicd10_flags$member) 

```

Use the next chunk to run smoking association with individual DNM rate (i.e., parentally phased DNMs) using **Model 6**.

```{r smoking x phased dnms, message=FALSE, warning=FALSE, results='asis'}

qc_covars = paste(colnames(phased_dnsnvs_wicd10_flags)[c(7:17)], collapse = "+")
smk_reg_formula=paste0("phased_dnsnvs~ever_smoked+age_at_offs+sex+sex*age_at_offs+", qc_covars) # This is Model 6

smk_reg = glm(data = phased_dnsnvs_wicd10_flags, 
    formula=smk_reg_formula, family="quasipoisson")


## Run regression by sex ## 
# Remove sex covariate and split dataset into females (1) and males (0)
smk_reg_bySex_formula = paste0("phased_dnsnvs~ever_smoked+age_at_offs+", qc_covars)

smk_reg_males_reg = glm(data = phased_dnsnvs_wicd10_flags %>% 
                          filter(sex==0), 
    formula=smk_reg_bySex_formula, family = "quasipoisson")

smk_reg_females_reg = glm(data = phased_dnsnvs_wicd10_flags %>% 
                            filter(sex==1), 
    formula=smk_reg_bySex_formula, family = "quasipoisson")

## Extract coefficients for the three models (sex concatenates, males only, females only)
sex_concat_smk_effect = summary(smk_reg)[["coefficients"]] %>%
  as.data.frame() %>%
  rownames_to_column(var = "term") %>%
  filter(term=="ever_smoked1") %>%
  mutate(sex_subset="both parents")

males_smk_effect = summary(smk_reg_males_reg)[["coefficients"]] %>%
  as.data.frame() %>%
  rownames_to_column(var = "term") %>%
  filter(term=="ever_smoked1") %>%
  mutate(sex_subset="fathers only")

females_smk_effect = summary(smk_reg_females_reg)[["coefficients"]] %>%
  as.data.frame() %>%
  rownames_to_column(var = "term") %>%
  filter(term=="ever_smoked1") %>%
  mutate(sex_subset="mothers only")

## concatenate effect estimate tables
all_ancestries_to_smoking = rbind(sex_concat_smk_effect, males_smk_effect, females_smk_effect) %>%
  dplyr::rename(se=`Std. Error`, estimate=Estimate, p=`Pr(>|t|)`) %>%
  mutate(model="all ancestries", 
         foldChange = exp(estimate),
         lci=exp(estimate-1.96*se), 
         uci=exp(estimate+1.96*se),
         significance_level=case_when(p<=0.0005 ~ "***", 
                             p<=0.005 ~ "**", 
                             p<=0.05 ~ "*"), 
         sex_subset=str_to_title(sex_subset), 
         sex_subset=factor(sex_subset, levels=c("Mothers Only", "Fathers Only", "Both Parents")))


```

See summary statistics corresponding to individual smoking effects (all_ancestries_to_smoking)

```{r smoking effects on phased dnms, echo=FALSE}

kable(all_ancestries_to_smoking)

```

Plot **main Figure 3** corresponding to DNM counts fold change differences between smokers (1) vs non-smokers (0). **NOTE:** a bug on rstudio/markdown prevents ploting of "significance_label" in the HTML render, but this works when running the code manually.

```{r plot smoking x phased dnms  fold changes, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}

# Significance label not plotting in HTML render! 

all_ancestries_to_smoking %>%
  ggplot(., aes(x=foldChange, y=sex_subset)) + 
  geom_errorbarh(aes(xmin = lci, xmax = uci), 
                position = position_dodge(0.9), 
                linewidth=6, height = 0, color="#99B898") +
  geom_point(size = 8, position = position_dodge(0.9), 
             color="black") + 
  geom_vline(xintercept = 1, linetype = "dotdash", 
             color = "black") + 
  geom_text(aes(label = significance_level), vjust=0, 
            show.legend = F, size=16, position = position_dodge(0.9), 
            fontface="bold", color="black") +
  labs(x = "Fold change", y="Parent subset") + 
  theme_classic(base_size = 20) +
  theme(strip.text.x = element_blank(),
    axis.title = element_text(size = 18, face="bold"),
    axis.text = element_text(size = 18, face="bold"),
    panel.grid.minor.y = element_blank(),
    legend.position = "none")


```
